compiler/rustc_mir_transform/src/sroa.rs

```diff
@@ -55,25 +56,24 @@ impl<'tcx> crate::MirPass<'tcx> for ScalarReplacementOfAggregates {
     }
 }
 
-/// Identify all locals that are not eligible for SROA.
-///
-/// There are 3 cases:
-/// - the aggregated local is used or passed to other code (function parameters and arguments);
-/// - the locals is a union or an enum;
-/// - the local's address is taken, and thus the relative addresses of the fields are observable to
-///   client code.
-fn escaping_locals<'tcx>(
+/// Also exclude locals whose types are ineligible for SRoA because of
+/// implementation restrictions (whether here or in codegen later).
+fn add_type_based_exclusions<'tcx>(
     tcx: TyCtxt<'tcx>,
-    excluded: &DenseBitSet<Local>,
+    excluded: &mut DenseBitSet<Local>,
     body: &Body<'tcx>,
-) -> DenseBitSet<Local> {
+) {
     let is_excluded_ty = |ty: Ty<'tcx>| {
```

Comment:
Why not just inline this closure? It's just a simple match on ADT, even if it's not written like that today.